### PR TITLE
Add visited website to memory for recalling content without being limited by the website summary.

### PR DESCRIPTION
**Pull request by maiko**
### Background
This pull request focus on improving the browsing functionality of Auto-GPT. 

The primary goal is to add chunk to the memory when browsing and summarizing websites, as well as to provide more configuration options to manage the chunk size and summary token count. The two new options are necessary to fine-tune the memory of chunk behavior. I had great result with smaller chunk max_lenght and bigger max_token for summary. 

The default for the new configuration options have been set to the current setting to avoid any unwanted change in behavior. 

### Changes

- Edited the config class in config.py to manage the browse_website command's chunk max_length, summary max_token count as well as better managed the User-Agent use for browsing website.
- Added memory management to browse.py to store raw content and summary in memory.
- Change the way User-Agent is handle in browse.py to be able to change it in the .env.
- Updated the commands.py file to pass along the url variable to the  browse.summarize_text function 

### Documentation
The changes are documented through in-code comments, and the configuration options have been added to the .env.template file. The new settings BROWSE_CHUNK_MAX_LENGTH and BROWSE_SUMMARY_MAX_TOKEN are explained in the config.py file, which also includes new methods for setting these values.

### Test Plan

- Configure the necessary environment variables in the .env file, including the new BROWSE_CHUNK_MAX_LENGTH and BROWSE_SUMMARY_MAX_TOKEN settings if you want to edit them.
- You can also change the User-Agent by setting USER_AGENT to a string corresponding to a full User-Agent string (see config.py). 
- Prompt Auto-GPT in a way to force her to execute a browse_website and to use memory. 

I've tested this PR using the following ai_settings.yaml:
```
ai_goals:
- 'Browse the web page: https://openai.com/product/gpt-4'
- 'Answer the question: What is the likeliness of GPT-4 to respond to requests for
  disallowed content ?'
- 'Answer the question: Is there a difference between GPT-3.5 and GPT-4 in producing
  factual responses ?'
- 'Asnwer the question: Provide two example of something built with GPT-4.'
- Use the task complete command to shut yourself down.
ai_name: WebsiteBrowserAI
ai_role: "an AI which only purpose is to browse a website and answer questions about it.\
         \/- You are NOT allowed to do Google searches to answer the questions.\
         \/- You are only allowed ***ONE*** use of the browse_website command.\
         \/- You ***CANNOT*** provide any of the question in the \"question\" argument when using the browse_website command.\
         \/- If unsure, thinking about a question will help you recall the answer."
```

As we will see below, I've asked Auto-GPT the answer to three questions. We will see that the summary provided by the browse_website doesn't include the information yet the AI will be able to answer the questions by "recalling" the answers from its memory. 

```
------------ CONTEXT SENT TO AI ---------------
System: The current time and date is Wed Apr 12 15:29:00 2023

System: This is an extract of your memory recalled by your thinking:
[]

User: Determine which next command to use, and respond using the format specified above:

----------- END OF CONTEXT ----------------
WEBSITEBROWSERAI THOUGHTS: I should start by browsing the webpage at https://openai.com/product/gpt-4 to find the answers to the questions provided. I will use the 'browse_website' co
REASONING: I need to navigate to the webpage and read about GPT-4 to answer the questions provided.
PLAN:
-  Use the browse_website command to navigate to https://openai.com/product/gpt-4
-  Read the information on the webpage to find answers to the questions provided
CRITICISM: I need to make sure I carefully read through the information on the webpage and not overlook any relevant information.
NEXT ACTION:  COMMAND = browse_website ARGUMENTS = {'url': 'https://openai.com/product/gpt-4', 'question': ''}
Enter 'y' to authorise command, 'y -N' to run N continuous commands, 'n' to exit program, or enter feedback for WebsiteBrowserAI...
Input:y
-=-=-=-=-=-=-= COMMAND AUTHORISED BY USER -=-=-=-=-=-=-=
Text length: 6692 characters
Adding chunk 1 / 2 to memory
Summarizing chunk 1 / 2
Added chunk 1 summary to memory
Adding chunk 2 / 2 to memory
Summarizing chunk 2 / 2
Added chunk 2 summary to memory
Summarized 2 chunks.
SYSTEM:  Command browse_website returned: Website Content Summary: "Result" : What is GPT-4 and what are its limitations? GPT-4 is OpenAI's most advanced system that can solve difficut has limitations such as social biases, hallucinations, and adversarial prompts, which OpenAI is addressing, and is working to encourage transparency, user education, and wider AI lids of text, surpassing ChatGPT in its advanced reasoning capabilities. Links: [' (/)', 'Skip to main content (/product/gpt-4#content)', ' (/)', 'Overview (/research/overview)', 'Index
Token limit: 4000

[...]
```
On the next round, based on what the AI is "thinking" (see the get_relevant_memory function), the context include chunk of website(s) relevant to the current tasks. 
```
------------ CONTEXT SENT TO AI ---------------
System: The current time and date is Wed Apr 12 15:29:34 2023

System: This is an extract of your memory recalled by your thinking:
["Source: https://openai.com/product/gpt-4\nContent summary part#2: What are the limitations of GPT-4?\n\nAccording to the text, GPT-4 still has many known limitations that are being I's efforts to encourage transparency, user education, and wider AI literacy as society adopts these models, and to expand the avenues of input people have in shaping their models.", ing: 12 pm - 12:30 pmGPT-4 outperforms ChatGPT by scoring in higher approximate percentiles among test-takers.Uniform Bar Exam10thChatGPT90thGPT-4Biology Olympiad31stChatGPT99thGPT-4 es more data and more computation to create increasingly sophisticated and capable language models.We spent 6 months making GPT-4 safer and more aligned. GPT-4 is 82% less likely to r on our internal evaluations.Safety & alignmentTraining with human feedbackWe incorporated more human feedback, including feedback submitted by ChatGPT users, to improve GPT-4â\x80\x9security.Continuous improvement from real-world useWeâ\x80\x99ve applied lessons from real-world use of our previous models into GPT-4â\x80\x99s safety research and monitoring system..GPT-4-assisted safety researchGPT-4â\x80\x99s advanced reasoning and instruction-following capabilities expedited our safety work. We used GPT-4 to help create training data for modeeâ\x80\x99ve collaborated with organizations building innovative products with GPT-4.DuolingoGPT-4 deepens the conversation on Duolingo.Be My EyesBe My Eyes uses GPT-4 to transform vitanleyMorgan Stanley wealth management deploys GPT-4 to organize its vast knowledge base.Khan AcademyKhan Academy explores the potential for GPT-4 in a limited pilot program.Government milestone in OpenAIâ\x80\x99s effort in scaling up deep learning.
[...]

WEBSITEBROWSERAI THOUGHTS: I should use the information from the website to answer the questions provided. Based on the information provided, GPT-4 is 82% less likely to respond to requests for disallowed content and 40% more likely to produce factual responses than GPT-3.5 on their internal evaluations. Therefore, I can use this information to answer the first question. I can use the do_nothing command to recall this information before answering the second question.
REASONING: Based on the information provided, GPT-4 is better in terms of producing factual responses and has improved safety standards as compared to GPT-3.5. I don't need to perform any actions to answer the first question, but I need to remember the information for the second question.
PLAN:
-  Use the information from the website to answer the first question
-  Use the do_nothing command to recall the information before answering the second question
CRITICISM: I need to ensure I recall the information accurately to answer the second question correctly.
NEXT ACTION:  COMMAND = do_nothing ARGUMENTS = {}
Enter 'y' to authorise command, 'y -N' to run N continuous commands, 'n' to exit program, or enter feedback for WebsiteBrowserAI...
Input:y
```
By "recalling" and getting relevant informations from memory, the AI is able to answer the subsequet question. In the next logs we can see that the included chunk is not the same as the more relevant for the context is now chunk 1 and was able to use that information to answer subsequent questions. 

```
------------ CONTEXT SENT TO AI ---------------
System: The current time and date is Wed Apr 12 15:30:09 2023

System: This is an extract of your memory recalled by your thinking:
["Source: https://openai.com/product/gpt-4\nContent summary part#1: What is GPT-4?\n\n

[...]
WEBSITEBROWSERAI THOUGHTS: Based on my previous recall of the information, and the examples built with GPT-4 mentioned on the website, I should answer the third question with the following examples: GPT-4 has been used to collaborate with organizations such as Duolingo to deepen conversation, Be My Eyes to transform visual accessibility, Stripe to streamline user experience and combat fraud, Morgan Stanley wealth management to organize its vast knowledge base, Khan Academy to explore the potential for GPT-4 and the government of Iceland to preserve its language.

```

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: scripts/config.py [hunk 1]
```diff
@@ -80,6 +83,7 @@ def __init__(self):
         # Note that indexes must be created on db 0 in redis, this is not configurable.
 
         self.memory_backend = os.getenv("MEMORY_BACKEND", 'local')
+
```
Comment (nponeccop) [inline scripts/config.py]: Rollback this unrelated whitespace change

File: scripts/main.py [hunk 2]
```diff
@@ -293,7 +293,7 @@ def parse_arguments():
     if args.gpt3only:
         print_to_console("GPT3.5 Only Mode: ", Fore.GREEN, "ENABLED")
         cfg.set_smart_llm_model(cfg.fast_llm_model)
-
```
Comment (nponeccop) [inline scripts/main.py]: Rollback this unrelated whitespace fix. We will fix them in a separate PR

**Review by nponeccop — CHANGES_REQUESTED** [review]

File: scripts/config.py (hunk 1 above)
Comment (maiko) [reply scripts/config.py]: done

File: scripts/main.py (hunk 2 above)
Comment (maiko) [reply scripts/main.py]: done

**Review by nponeccop — DISMISSED** [review]

**Comment by maiko** [comment]
There was two conflict that I just resolved.

File: .env.template [hunk 3]
```diff
@@ -17,4 +17,6 @@ OPENAI_AZURE_EMBEDDINGS_DEPLOYMENT_ID=deployment-id-for-azure-embeddigs
 IMAGE_PROVIDER=dalle
 HUGGINGFACE_API_TOKEN=
 USE_MAC_OS_TTS=False
-MEMORY_BACKEND=local
+BROWSE_CHUNK_MAX_LENGTH=4000
+BROWSE_SUMMARY_MAX_TOKEN=300
+MEMORY_BACKEND=local
```
Comment (nponeccop) [inline .env.template]: Add the final line terminator back. You deleted it by mistake

**Review by nponeccop — CHANGES_REQUESTED** [review]

**Review by nponeccop — APPROVED** [review]

**Comment by nponeccop** [comment]
@maiko There are conflicts

**Comment by maiko** [comment]
And again, conflict resolved. 
This need to be merge as to avoid any new commit. 

@Torantulino Can you have a look and merge is happy ? I do believe this will be a great addition to Auto-GPT.

**Review by nponeccop — APPROVED** [review]

**Comment by nponeccop** [comment]
Ask a review from me to speed it up. Or tag me. I'm the first one in the current merging pipeline.

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@maiko There are conflicts now

**Comment by maiko** [comment]
Conflicts resolved and I've used the new .env.template presentation.

**Review by nponeccop — APPROVED** [review]

File: .env.template [hunk 4]
```diff
@@ -3,6 +3,12 @@
 ################################################################################
 # EXECUTE_LOCAL_COMMANDS - Allow local command execution (Example: False)
 EXECUTE_LOCAL_COMMANDS=False
+# BROWSE_CHUNK_MAX_LENGTH - When browsing website, define the length of chunk stored in memory
+BROWSE_CHUNK_MAX_LENGTH=4000
```
Comment (richbeales) [inline .env.template]: this should match the value you've given in config.py

**Comment by maiko** [comment]
default value are now aligned between .env.template and config.

**Review by nponeccop — DISMISSED** [review]

**Comment by richbeales** [comment]
Run flake8 scripts/ tests/ --select E303,W293,W291,W292,E305,E231,E302
scripts/browse.py:140:1: E302 expected 2 blank lines, found 1
Error: Process completed with exit code 1.

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@maiko There are conflicts again due to merging.

**Comment by maiko** [comment]
and resolved

**Review by richbeales — APPROVED** [review]
